src/main/java/com/google/vinet/data/TimeSlot.java

```diff
@@ -0,0 +1,128 @@
+/*
+ *  Copyright 2020 Google LLC
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      https:www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.google.vinet.data;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+public abstract class TimeSlot implements Comparable<TimeSlot> {
+
+  private final Instant start;
+  private final Instant end;
+  protected final RegisteredUser registeredUser;
+  private final Set<TimeSlot> neighbours = new HashSet<>();
+  private TimeSlot pairedSlot = null;
+  private double distance = 0;
+
+  /**
+   * Create a new time slot. Start and end times must not be null, and start time must come before
+   * end time.
+   *
+   * @param start          The starting instant for the slot.
+   * @param end            The ending instant for the slot.
+   * @param registeredUser The user for which this slot applies.
+   */
+  public TimeSlot(Instant start, Instant end, RegisteredUser registeredUser) {
+    this.start = Objects.requireNonNull(start, "Start time must not be null!");
+    this.end = Objects.requireNonNull(end, "End time must not be null!");
+    if (start.isAfter(end)) throw new IllegalArgumentException();
```

Comment:
Line 46: Minor logical error I think, `start.isAfter(end)` doesn't ensure the requirement that 'start time must come before end time'. `if ( !start.isBefore(end) )` might work instead

Reply:
Is that not the same thing? Possibly excluding the case where the start and end are the same.

Reply:
Yes it's the case of start and end being the same that would be allowed, though it shouldn't be. Any of `!start.isBefore(end)`, `!end.isAfter(start)`, `start.isAfter(end) || start.equals(end)`  would work I think.

Reply:
Ok this has now been change to prevent the start and end times being the same.